src/services/lnd/index.ts

```diff
@@ -623,3 +625,51 @@ const translateLnPaymentLookup = (p): LnPaymentLookup => ({
     : undefined,
   attempts: p.attempts,
 })
+
+const resolvePaymentStatus = async ({
+  lnd,
+  result,
+}: {
+  lnd: AuthenticatedLnd
+  result: GetPaymentResult
+}): Promise<PaymentStatus> => {
+  const { is_confirmed, is_failed, payment, pending } = result
+
+  if (is_confirmed) return PaymentStatus.Settled
+  if (is_failed) return PaymentStatus.Failed
+
+  const cache = LocalCacheService()
+  const currentBlockHeight = await cache.getOrSet({
+    key: CacheKeys.BlockHeight,
+    ttlSecs: SECS_PER_5_MINS,
+    fn: async () => {
+      const { current_block_height } = await getWalletInfo({ lnd })
+      return current_block_height
+    },
+  })
+
+  const timeout = pending?.timeout || payment?.timeout
+  // This is a hack to handle lnd issue with pending HTLCs after channel close
+  // https://github.com/lightningnetwork/lnd/issues/6249
+  if (timeout && pending && currentBlockHeight > timeout) {
+    const closedChannels = await cache.getOrSet({
+      key: CacheKeys.ClosedChannels,
+      ttlSecs: SECS_PER_5_MINS,
+      fn: async () => {
+        const { channels } = await getClosedChannels({ lnd })
+        return channels
+      },
+    })
+
+    const failed = pending.paths
+      .map((p) => {
+        const [first] = p.hops
+        return closedChannels.find((c) => c.id === first.channel)
+      })
+      .every((s) => !!s)
+
+    if (failed) return PaymentStatus.Failed
```

Comment:
do we have all this logic just to return `PaymentStatus.Failed` instead of `PaymentStatus.Pending`?

shouldn't we focus on making sure payment can't be pending after some time, and settle it properly in the ledger?

I dont think this approach is the right one, because we may ends up with many pending that could be pending forever in the ledger, and will have to go through this loop forever because they have not been settled properly. 

this will, over time, introduce more and more discrepancy in our ledger.

Reply:
do we have all this logic just to return PaymentStatus.Failed instead of PaymentStatus.Pending? **yes**

shouldn't we focus on making sure payment can't be pending after some time, and settle it properly in the ledger? **this method is used by updatePendingPayments so it will update ledger when return failed. I think is a clean and easy way to remove it later if lnd is fixed**

Reply:
ok that make sense then. thanks for the clarification.